base/shared/src/main/scala/scalaz/data/Freer.scala

```diff
@@ -0,0 +1,86 @@
+package scalaz
+package data
+
+import scala.annotation.tailrec
+
+/**
+  * Free(r) Monad - arrows A => Free[F, B] are concatenated using a type-aligned
+  * stack-ended queue ([[ACatenable1]]), for better complexity on certain types
+  * of Free monad use cases.
+  *
+  * For more information, [see Ed Kmett's comments here.](https://www.reddit.com/r/haskell/comments/7q4sku/are_people_using_freer_monads_or_still_mostly/dsmlnh7/)
+  *
+  */
+sealed abstract class Freer[F[_], A] {
+
+  def map[B](f: A => B): Freer[F, B] =
+    flatMap[B](a => Freer.pure[F, B](f(a)))
+
+  def flatMap[B](k: A => Freer[F, B]): Freer[F, B] =
+    this match {
+      case Freer.Pure(a) => k(a)
+      case l @ Freer.LiftF(_) => Freer.impure(l, Freer.singleK(Kleisli.wrapKleisli(k)))
+      case Freer.Impure(ff, rs) => Freer.impure(ff, rs :+ Kleisli.wrapKleisli(k))
+    }
+
+  final def hoist[G[_]](α: F ~> Freer[G, ?]): Freer[G, A] =
+    foldMap(α)(Freer.monad)
+
+  final def foldFreer[B](la: F[Freer[F, A]] => B)(ra: A => B)(implicit F: Functor[F]): B =
+    runFreer.fold(la)(ra)
+
+  final def foldMap[M[_]](α: F ~> M)(implicit M: Monad[M]): M[A] =
+    step match {
+      case Freer.Pure(a) => M.applicative.pure(a)
+      case Freer.LiftF(fa) => Forall.specialize[λ[α => F[α] => M[α]], A](α).apply(fa)
+      case Freer.Impure(f, q) => M.bind.flatMap(f foldMap α)(a => Freer.runQuiver(f)(q).foldMap(α))
+    }
+
+  @tailrec
+  final def runFreer(implicit F: Functor[F]): F[Freer[F, A]] \/ A =
+    this match {
+      case Freer.Pure(a) => \/-(a)
+      case Freer.LiftF(fa) => -\/(F.map(fa)(Freer.pure))
+      case Freer.Impure(f, q) =>
+        f match {
+          case Freer.Pure(b) => Kleisli.runKleisli(q.fold(KleisliImpl.kleisliCompose)).apply(b).runFreer
+          case Freer.LiftF(fa) => -\/(F.map(fa)(Kleisli.runKleisli(q.fold(KleisliImpl.kleisliCompose))))
+          case Freer.Impure(ff, qq) => Freer.runQuiver(ff)(qq >>> q).runFreer
+        }
+    }
+
+  @tailrec
+  private[Freer] final def step: Freer[F, A] =
+    this match {
+      case Freer.Impure(f, q) => Freer.runQuiver(f)(q).step
+      case f => f
+    }
+}
+
+object Freer extends FreerInstances {
+  private[Freer] type =>:[F[_], A, B] = Kleisli[Freer[F, ?], A, B]
+  private[Freer] type Quiver[F[_], A, B] = ACatenable1[=>:[F, ?, ?], A, B]
+
+  final case class Pure[F[_], A] private (a: A) extends Freer[F, A]
+  final case class LiftF[F[_], A] private (fa: F[A]) extends Freer[F, A]
+  final case class Impure[F[_], E, A] private (f: Freer[F, E], q: Quiver[F, E, A]) extends Freer[F, A] {
```

Comment:
I don't think we can make any use of the extra reflection here (a list of functions is as opaque as a single function), and the extra level of indirection will only make it slower than your other `Free` encoding.